media_cdn/dualtoken.rb

```diff
@@ -0,0 +1,170 @@
+        """
+        Gets the Signed URL Suffix string for the Media CDN' Short token URL requests.
+        One of (`url_prefix`, `full_path`, `path_globs`) must be included in each input.
+        Args:
+            base64_key: Secret key as a base64 encoded string.
+            signature_algorithm: Algorithm can be either `SHA1` or `SHA256` or `Ed25519`.
+            start_time: Start time as a UTC datetime object.
+            expiration_time: Expiration time as a UTC datetime object. If None, an expiration time 5 mins from now will be used.
+            url_prefix: the URL prefix to sign, including protocol.
+                        For example: http://example.com/path/ for URLs under /path or http://example.com/path?param=1
+            full_path:  A full path to sign, starting with the first '/'.
+                        For example: /path/to/content.mp4
+            path_globs: a set of ','- or '!'-delimited path glob strings.
+                        For example: /tv/*!/film/* to sign paths starting with /tv/ or /film/ in any URL.
+            session_id: a unique identifier for the session
+            data: data payload to include in the token
+            headers: header name and value to include in the signed token in name=value format.  May be specified more than once.
+                        For example: [{'name': 'foo', 'value': 'bar'}, {'name': 'baz', 'value': 'qux'}]
+            ip_ranges: A list of comma separated ip ranges. Both IPv4 and IPv6 ranges are acceptable.
+                        For example: '203.0.113.0/24,2001:db8:4a7f:a732/64'
+
+        Returns:
+            The Signed URL appended with the query parameters based on the
+            specified URL prefix and configuration.
+        """
+
+        decoded_key = Base64.urlsafe_decode64(base64_key)
+        algo = signature_algorithm.downcase
+
+        # For most fields, the value we put in the token and the value we must sign
+        # are the same.  The FullPath and Headers use a different string for the
+        # value to be signed compared to the token.  To illustrate this difference,
+        # we'll keep the token and the value to be signed separate.
+        tokens = Array.new
+        to_sign = Array.new
+
+        # check for `full_path` or `path_globs` or `url_prefix`
+        if full_path
+            tokens.append("FullPath")
+            to_sign.append("FullPath=#{full_path}")
+        elsif path_globs
+            path_globs = path_globs.strip
+            field = "PathGlobs=#{path_globs}"
+            tokens.append(field)
+            to_sign.append(field)
+        elsif url_prefix
+            field = "URLPrefix=" + base64_encoder(url_prefix).encode("utf-8")
+            tokens.append(field)
+            to_sign.append(field)
+        else
+            raise "User Input Missing: One of `url_prefix`, `full_path` or `path_globs` must be specified"
```

Comment:
Make the case and punctuation changes suggested here: "User input missing: One of `url_prefix`, `full_path`, or `path_globs` must be specified."